src/app/lib/business/dad_jokes/dad_jokes_service.dart

```diff
@@ -0,0 +1,164 @@
+import 'dart:convert';
+
+import 'package:app/access/dad_jokes/dad_jokes_repository.dart';
+import 'package:app/access/dad_jokes/data/dad_joke_content_data.dart';
+import 'package:app/business/dad_jokes/dad_joke.dart';
+import 'package:get_it/get_it.dart';
+import 'package:rxdart/rxdart.dart';
+import 'package:shared_preferences/shared_preferences.dart';
+
+/// Interface for the dad jokes service.
+abstract interface class DadJokesService implements Disposable {
+  factory DadJokesService(DadJokesRepository dadJokesRepository) =
+      _DadJokesService;
+
+  /// Stream of dad jokes.
+  Stream<List<DadJoke>> get dadJokesStream;
+
+  /// Get dad jokes.
+  Future<List<DadJoke>> getDadJokes();
+
+  /// Get favorite dad jokes.
+  Future<List<DadJoke>> getFavoriteDadJokes();
+
+  /// Add a favorite dad joke.
+  Future addFavoriteDadJoke(DadJoke joke);
+
+  /// Remove a favorite dad joke.
+  Future removeFavoriteDadJoke(DadJoke joke);
+}
+
+/// Implementation of [DadJokesService].
+final class _DadJokesService implements DadJokesService {
+  /// The key used to store favorite dad jokes in shared preferences.
+  final String _favoriteDadJokesKey = 'favoriteDadJokes';
+
+  /// The behavior subject used to notify when dad jokes changed.
+  final BehaviorSubject<List<DadJoke>> _dadJokesBehaviorSubject =
+      BehaviorSubject();
+
+  /// The repository used to fetch dad jokes.
+  final DadJokesRepository _dadJokesRepository;
+
+  _DadJokesService(DadJokesRepository dadJokesRepository)
+      : _dadJokesRepository = dadJokesRepository {
+    getDadJokes().then((dadJokes) => _dadJokesBehaviorSubject.add(dadJokes));
+  }
+
+  @override
+  Stream<List<DadJoke>> get dadJokesStream => _dadJokesBehaviorSubject.stream;
+
+  @override
+  Future<List<DadJoke>> getDadJokes() async {
+    final dadJokeResponseData = await _dadJokesRepository.getDadJokes();
+    final favoriteDadJokes = await getFavoriteDadJokes();
+
+    return dadJokeResponseData.dadJokeData.dadJokeChildrenData.map(
+      (dadJokeChildData) {
+        return DadJoke(
+          id: dadJokeChildData.dadJokeContentData.id,
+          title: dadJokeChildData.dadJokeContentData.title,
+          selfText: dadJokeChildData.dadJokeContentData.selfText,
+          isFavorite: favoriteDadJokes.any(
+            (favoriteDadJoke) =>
+                favoriteDadJoke.id == dadJokeChildData.dadJokeContentData.id,
+          ),
+        );
+      },
+    ).toList();
+  }
+
+  @override
+  Future<List<DadJoke>> getFavoriteDadJokes() async {
+    final favoriteDadJokesJson = await _getFavoriteDadJokesJson();
+
+    return favoriteDadJokesJson
+        .map(
+          (favoriteDadJokeJson) => DadJoke.fromData(
+            DadJokeContentData.fromJson(jsonDecode(favoriteDadJokeJson)),
+            isFavorite: true,
+          ),
+        )
+        .toList();
+  }
+
+  @override
+  Future addFavoriteDadJoke(DadJoke dadJoke) async {
+    final sharedPreferences = await SharedPreferences.getInstance();
```

Comment:
Your business service should not be coupled with a data source. It should rely on repositories (as abstractions).